src/boxrobots/src/controllers/QPinnerloopcontroller.jl

```diff
@@ -0,0 +1,476 @@
+immutable NoSwitch <: ContactSwitchType end
+immutable MakingContact <: ContactSwitchType end
+immutable BreakingContact <: ContactSwitchType end
+
+type ContactSwitch{T<:ContactSwitchType}
+  limb_symbol::Symbol
+end
+
+function ContactSwitch{T}(limb_symbol::Symbol, ::Type{T})
+  return ContactSwitch{T}(limb_symbol)
+end
+
+function ContactSwitch(cs_0::ContactState, cs_1::ContactState)
+  """
+  Compute the contact switch type when transitioning from
+  cs_0 --> cs_1
+  """
+
+  num_contact_state_mismatches = 0
+  limb_symbol = Symbol()
+  contact_switch_type = ContactSwitchType
+
+  for (limb_sym, contact_indicator_0) in cs_0
+    contact_indicator_1 = cs_1[limb_sym]
+    # there was a contact switch
+    if contact_indicator_0 != contact_indicator_1
+      if (contact_indicator_0 & !contact_indicator_1)
+        contact_switch_type = BreakingContact
+      else
+        contact_switch_type = MakingContact
+      end
+      limb_symbol = limb_sym
+      num_contact_state_mismatches += 1
+    end
+  end
+
+  if num_contact_state_mismatches > 1
+    error("more than one state changed simultaneously, dont know how to handle this")
+  end
+
+  if num_contact_state_mismatches > 0
+    contact_switch = ContactSwitch(limb_symbol, contact_switch_type)
+    return true, contact_switch
+  else
+    return false, nothing
+  end
+
+end
+
+function get_contact_switch_type{T}(contact_switch::ContactSwitch{T})
+  return T
+end
+
+
+abstract ContactMismatchType
+immutable NoMismatch <: ContactMismatchType end
+immutable EarlyContact <: ContactMismatchType end
+immutable LateContact <: ContactMismatchType end
+
+type ContactMismatch{T<:ContactMismatchType}
+  limb_sym::Symbol
+end
+
+function ContactMismatch{T}(limb_symbol::Symbol, ::Type{T})
+  return ContactMismatch{T}(limb_symbol)
+end
+
+type QPInnerLoopControllerState
+  t_start::Float64 # time at which soln_data was constructed
+  soln_data::PyObject
+  contact_plan::ContactPlan
+  contact_state::ContactState
+  contact_mismatch::ContactMismatch
+end
+
+function QPInnerLoopControllerState(t_start::Float64, state::BoxRobotState, soln_data::PyObject)
+  contact_state = contact_state_from_robot_state(state)
+  contact_plan = make_contact_plan(contact_state, soln_data)
+  contact_mismatch = ContactMismatch(Symbol(), NoMismatch)
+  return QPInnerLoopControllerState(t_start, soln_data, contact_plan, contact_state,
+  contact_mismatch)
+end
+
+type QPInnerLoopController <: BoxRobotController
+  python_controller::PyObject
+  controller_state::QPInnerLoopControllerState
+end
+
+function QPInnerLoopController(python_controller::PyObject, t_start::Float64,
+  state::BoxRobotState, soln_data::PyObject)
+  controller_state = QPInnerLoopControllerState(t_start, state, soln_data)
+  return QPInnerLoopController(python_controller, controller_state)
+end
+
+type QPInnerLoopControllerData <: BoxRobotControllerData{QPInnerLoopController}
+  solve_time::Float64
+  contact_assignments::Dict{Symbol, Vector}
```

Comment:
Another case where you might want a more strict type: `Vector` is a parametric type, so if you want faster code, you'll need to specify its element type like

```julia
type Data{T}
  assignments::Dict{Symbol, Vector{T}}
end
```